tests/infer/test_valid_models.py

```diff
@@ -2344,6 +2344,58 @@ def guide():
     assert_ok(model, guide, elbo)
 
 
+@pytest.mark.stage("funsor")
+@pytest.mark.parametrize("num_particles", [1, 2])
+def test_collapse_diag_normal_plate_normal(num_particles):
+    T, d = 5, 3
+    data = torch.randn((T, d))
+
+    def model():
+        x = pyro.sample("x", dist.Normal(0., 1.)).unsqueeze(-1)
+        with poutine.collapse():
+            with pyro.plate("data", T, dim=-1):
+                expand_shape = (d,) if num_particles == 1 else (num_particles, 1, d)
+                y = pyro.sample("y", dist.Normal(x, 1.).expand(expand_shape).to_event(1))
+                pyro.sample("z", dist.Normal(y, 1.).to_event(1), obs=data)
+
+    def guide():
+        loc = pyro.param("loc", torch.tensor(0.))
+        scale = pyro.param("scale", torch.tensor(1.), constraint=constraints.positive)
+        pyro.sample("x", dist.Normal(loc, scale))
+
+    elbo = Trace_ELBO(num_particles=num_particles, vectorize_particles=True,
+                      max_plate_nesting=1)
+    assert_ok(model, guide, elbo)
+
+
+@pytest.mark.stage("funsor")
+@pytest.mark.parametrize("num_particles", [1, 2])
+def test_collapse_normal_mvn_mvn(num_particles):
+    T, d, S = 5, 3, 4
+    data = torch.randn((T, d, S))
+
+    def model():
+        x = pyro.sample("x", dist.Normal(0., 1.)).unsqueeze(-1)
+        with poutine.collapse():
+            with pyro.plate("d", d, dim=-1):
+                expand_shape = (d, S) if num_particles == 1 else (num_particles, d, S)
+                beta0 = pyro.sample("beta0", dist.Normal(x, 1.).expand(expand_shape).to_event(1))
+                beta = pyro.sample("beta", dist.MultivariateNormal(beta0, torch.eye(S)))
+
+            mean = torch.ones((T, d)) @ beta
```

Comment:
This fails because `beta.output` is `Reals[S]` while we need it to be `Reals[d, S]`.

Reply:
Yeah this seems like a major incompatibility between funsor-style operations which have a clear batch/event split and numpy-style operations where everything is effectively an event dim. In particular there's no way for `ops.matmul` to touch the batch dimension "d".

One workaround in this model would be to move `beta0` and `beta` out of the plate and instead use `.to_event(1)`, and I think that kind of workaround will be needed whenever we exit a plate and treat a formerly-batch dimension as an event dimension. Conversely I conjecture that no workarounds are needed in tree-plated models, that is in models where "no variable outside a plate ever depends on an upstream variable inside a plate"; this class is similar to our condition for tractable tensor variable elimination.

Reply:
It seems like if we had a way of knowing the plate context of a value each time it was accessed, we could handle this smoothly using `funsor.terms.Lambda`. Suppose we could overload variable access/assignment, e.g. by using an overloadable environment data structure `env` rather than `locals()`:
```py
with plate("plate_var", d, dim=-1):
    beta0 = pyro.sample("beta0", dist.Normal(x, 1.).expand(expand_shape).to_event(1))
    # setting env.beta also records the current plate context of beta
    env.beta = pyro.sample("beta", dist.MultivariateNormal(beta0, torch.eye(S)))

# use funsor.terms.Lambda to convert the dead plate dimension to an output:
# now reading env.beta returns funsor.Lambda(plate_var, beta)
# where plate_var is the difference between the current and original plate contexts
mean = torch.ones((T, d)) @ env.beta
```

Reply:
Oh, I didn't know about `Lambda`. I am not sure why we need `env` but it seems from your code that in the `collapse` code, when the output does not match `plate` infos, we can use `Lambda`.

Reply:
> I am not sure why we need env

I was suggesting it as a layer of automation on top of `Lambda` that would help keep track of which plate dimensions need to be converted to event dimensions via `Lambda`, just like `contrib.funsor.to_funsor`/`to_data` automatically track the `name_to_dim` mapping for enumeration.  You're right that it's also possible to use `Lambda` directly in user code.